fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java

```diff
@@ -201,4 +209,67 @@ public ScalarOperator optimizeDateTimeToDateCast(BinaryPredicateOperator operato
+        LocalDateTime targetDate;
+        BinaryPredicateOperator.BinaryType binaryType = operator.getBinaryType();
+        int offset;
+        BinaryPredicateOperator resultBinaryPredicateOperator;
+        ConstantOperator newDatetime;
+        switch (binaryType) {
+            case GE:
+                // when the BinaryType is >= , cast date to equivalent datetime type
+                // E.g. cast(id_datetime as date) >= 2021-12-28
+                // optimized to id_datetime >= 2021-12-28 00:00:00.0
+                offset = 0;
+                targetDate = originalDate.plusDays(offset);
+                newDatetime = ConstantOperator.createDatetime(targetDate.toLocalDate().atTime(0, 0, 0, 0));
+                resultBinaryPredicateOperator =
+                        new BinaryPredicateOperator(BinaryPredicateOperator.BinaryType.GE, castChild, newDatetime);
+                break;
+            case GT:
+                // when the BinaryType is > , cast date to equivalent datetime type of next day
+                // E.g. cast(id_datetime as date) > 2021-12-28
+                // optimized to id_datetime >= 2021-12-29 00:00:00.0
+                offset = 1;
+                targetDate = originalDate.plusDays(offset);
+                newDatetime = ConstantOperator.createDatetime(targetDate.toLocalDate().atTime(0, 0, 0, 0));
+                resultBinaryPredicateOperator =
+                        new BinaryPredicateOperator(BinaryPredicateOperator.BinaryType.GE, castChild, newDatetime);
+                break;
+            case LE:
+                // when the BinaryType is <= , cast date to equivalent datetime type of next day
+                // E.g. cast(id_datetime as date) <= 2021-12-28
+                // optimized to id_datetime < 2021-12-29 00:00:00.0
+                offset = 1;
+                targetDate = originalDate.plusDays(offset);
+                newDatetime = ConstantOperator.createDatetime(targetDate.toLocalDate().atTime(0, 0, 0, 0));
+                resultBinaryPredicateOperator =
+                        new BinaryPredicateOperator(BinaryPredicateOperator.BinaryType.LT, castChild, newDatetime);
+                break;
+            case LT:
+                // when the BinaryType is < , cast date to equivalent datetime type
+                // E.g. cast(id_datetime as date) < 2021-12-28
+                // optimized to id_datetime < 2021-12-28 00:00:00.0
+                offset = 0;
+                targetDate = originalDate.plusDays(offset);
+                newDatetime = ConstantOperator.createDatetime(targetDate.toLocalDate().atTime(0, 0, 0, 0));
+                resultBinaryPredicateOperator =
+                        new BinaryPredicateOperator(BinaryPredicateOperator.BinaryType.LT, castChild, newDatetime);
+                break;
+            case EQ:
+                // Do not optimize when the BinaryType is =, otherwise we need replace it with BetweenPredicateOperator
```

Comment:
I think trans to BetweenPredicateOperator may be better for BE to scan because of the prefix index

Reply:
+1

Reply:
Ok